torch/_inductor/codecache.py

```diff
@@ -821,21 +821,27 @@ def _save_graph(
         write_atomic(path, content)
 
     @staticmethod
-    def _check_can_cache():
+    def _check_can_cache(gm: torch.fx.GraphModule):
         """
         Check some conditions that would preclude caching and raise BypassFxGraphCache
         to bypass in case caching is not possible.
         """
+        # Freezing can embed constants that wouldn't be static across runs.
         if config.freezing or config.aot_inductor.use_runtime_constant_folding:
-            # Freezing can embed constants that wouldn't be static across runs.
             raise BypassFxGraphCache()
 
+        # The treatment of guards in the caching implementation requires that
+        # we have a shape env.
         if FxGraphCache._get_shape_env() is None:
-            # The treatment of guards in the caching implementation requires that
-            # we have a shape env.
             log.debug("fx graph cache no shape env")
             raise BypassFxGraphCache()
 
+        # HigherOrderOperators should be handled on a case-by-case basis.
+        # Currently, we just skip caching if we have any.
```

Comment:
should we be skipping all of them by default? i assume cond/map etc should just work out of the box

Reply:
This was @eellison's suggestion. Elias, what do you think? What is the set of higher order ops? Is it currently just triton + cond/map, or are there others as well? And if there are others, is there a criteria to distinguish which we expect to work vs. not?

Reply:
A subset of all higher order ops: https://fburl.com/biggrep/h06kfrgg

Reply:
@oulgen Ah thanks. So there are several. So do you think I should allowlist cond/map here specifically, i.e., because they're important? Or were those just two semi-random examples and you're advocating to allow more?

Reply:
I suspect they are more popular HOPs but perhaps @zou3519 would have a more definitive answer on what is safe here

Reply:
If I'm understanding correctly, we cache compiled FX graph from inductor. This PR is saying that a graph that contains HOPs cannot be cached by default.

If so, that sounds fine to me -- caching for things like torch.cond get a bit complicated because you also want to walk through the subgraphs. cc @aakhundov

Reply:
@zou3519 so Sam added a cache from FX graph to inductor generated python file. This does not work for triton HOPs because the idx on the dynamo side table needs to match which is tricky. But for torch.cond etc everything is on the graph, so I think that should be fine to cache.

Reply:
@zou3519 

> Is the cache from FX GraphModule to inductor generated python file?

Exactly.

> This PR is saying that a graph that contains HOPs cannot be cached by default.

Correct. In other words, if we find find any HOPs, then bypass the cache altogether and compile normally.

> And do we recurse into the submodules of the FX GraphModule?

Hmmm, I guess I still don't know enough about inductor compilation to answer that. Maybe @eellison can help give a definite answer. Is the question about how we compute the key to look up entries or the makeup of the cache entries themselves? I don't know if a code pointer will help, but this is where the cache plugs in: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L464-L471. So at a high level, we compute the cache key from the FX GraphModule (the string representation of the code actually) and the inputs, and then we cache the output of `fx_codegen_and_compile()`

Reply:
torch.cond and torch.while could be supported, but they dont work "out of the box". we'd need to recurse into the subgraphs. maybe we can do that automatically, but it does not work as is today. I think it's fine to add safety in this pr and follow up and add support. 

It could also make more sense for folks who are adding the HOP to add the cache key then sam - they will know better about the details of what needs to be cached.